doc/ecs/create_own_content.md

```diff
@@ -0,0 +1,82 @@
+## Entitäten erstellen
+
+Anders als im klassischen OOP werden Entitäten nicht durch das Ableiten der `Entity` Klasse erzeugt und definiert, sondern durch das Hinzufügen von Komponente.
+
+Um eine eigene Entität zu erstellen, muss man:
+
+1. Eine neue Instanz vom Typ `Entity` anlegen
+2. Die gewünschten Komponenten erstellen
+3. Die Komponenten mit `Entity#addComponent` hinzufügen
+
+Beispiel: 
+```java
+Entity monster = new Entity();
+PositionComponent pc= new PositionComponent(monster);
+AIComponent ai = new AIComponent(monster);
+AnimationComponent ac = new AnimationComponent(monster);
+VelocityComponent vc = new AnimationComponent(monster);
+
+monster.addComponent(PositionComponent.name,pc);
+monster.addComponent(AIComponent.name,ai);
+monster.addComponent(AnimationComponent.name,ac);
+monster.addComponent(VelocityComponent.name,vc);
+```
+
+*Anmerkung: Die Komponenten wurden im Beispiel alle mit den jeweiligen Default-Werten initialisiert. Die Verwendung der anderen Konstruktoren, mit eigenen Werten, geht natürlich auch.* 
+
+## Component erstellen
+Um eigene Components zu implementieren, muss die eigene Klasse von der abstrakten Klasse  `Component` abgeleitet werden.
+
+Components werden im package `ecs.components` abgelegt und sollen den Namensschema `$WHAT_IS_THIS_COMPONENT$Component`folgen.
+Jedes Component braucht eine statische Variable `String name`, um den Component-Typen zu identifizieren. Dieser Name wird als Key-Value des HashSets für die, in den Entitäten gespeicherten, Components verwendet. 
+
+Jede Component-Instanz gehört zu genau einer Entitäs-Instanz. Eine Entitäts-Instanz kann einen Component-Typen nur einmal speichern.
+
+Um Components für die DSL verfügbar zu machen, siehe TBD
+
+## System erstellen
+Um eigenen Systeme zu implementieren, muss die eigene Klasse von `ECS_System` abgeleitet werden.
+
+Die Funktionalität des Systems wird in der `update`-Methode implementiert. Diese wird einmal pro Frame aufgerufen.
+
+In der `update`-Methode wird dann über das HashSet `ECS.entities`iteriert.
+Dabei sollte zuerst geprüft werden, ob die Entität das "Key"-Component enthält. 
+Das "Key"-Component ist das Component, welches dem System sagen soll, dass die Entität unter dem System fällt, die Logik des Systems also auf die Entität angewendet werden soll.
+Beispiel: Das `VelocitySystem` soll nur auf Entitäten mit dem `VelocityComponent` agieren.
+
+Beispiel aus dem `VelocitySystem`: 
+```java
+ for (Entity entity : ECS.entities) {
```

Comment:
```java
ECS.entities.stream()
            .flatMap(entity -> entity.getComponent(VelocityComponent.name).stream())
            .forEach(optcomponent -> <own code>);
```

Nicht getestet, aber von der Idee her die konsequentere Umsetzung von/mit `Option`.

- Mache aus dem Set einen Stream von Entities
- Gehe über jedes Entity und hole die Component mit dem Namen "VelocityComponent.name"
    - Technisch ist das Ergebnis ein `Option<Component>`
    - Mache aus diesem Option-Component einen kleinen `Stream<Component>`: leer, falls kein Element im Option-Component; sonst das ausgepackte Element (die gesuchte Component) im Stream
    - `flatMap` mergt die ganzen Teil-Streams zu einem einzigen `Stream<Component>`
- `forEach` ist eine Art `map`, das eine terminale Operation darstellt und den Stream abschließt. Sonst müsste man `map` nehmen und den Stream irgendwie mit `count()` oder so terminieren (jaja, Java ist Mist).

Reply:
Na gut, man sucht ja auch nach bestimmten Components ... dann vielleicht doch nicht in der Basisklasse.

Reply:
Eben kurz getestet, der Code funktioniert tatsächlich so. Die beiden Schnipsel sind äquivalent: 

```java
        for (Entity entity : ECS.entities) {
            entity.getComponent("wuppie").ifPresent(
                    vc -> {
                        System.out.println(vc);
                    });
        }

        ECS.entities.stream()
                .flatMap(entity -> entity.getComponent("wuppie").stream())
                .forEach(System.out::println);
```

Reply:
Man könnte das auch in drei Schritten schreiben, dann ist es evtl. für Ungeübte lesbarer:

```java
        ECS.entities.stream()
                .map(entity -> entity.getComponent("wuppie"))
                .filter(Optional::isPresent)
                .forEach(System.out::println);
```